File: crates/runtime/src/builtin/node_http_disabled.js
// node:http stub when http feature is disabled

const NOT_SUPPORTED_ERROR = new Error('node:http requires the "http" feature flag to be enabled');

export const METHODS = [];
export const STATUS_CODES = {};
export const maxHeaderSize = 16384;

export function validateHeaderName() {}
export function validateHeaderValue() {}

export class Agent {
    constructor() { throw NOT_SUPPORTED_ERROR; }
}

export const globalAgent = null;

export class ClientRequest {
    constructor() { throw NOT_SUPPORTED_ERROR; }
}

export class IncomingMessage {
    constructor() { throw NOT_SUPPORTED_ERROR; }
}

export class Server {
    constructor() { throw NOT_SUPPORTED_ERROR; }
}

export class ServerResponse {
    constructor() { throw NOT_SUPPORTED_ERROR; }
}

export function createServer() { throw NOT_SUPPORTED_ERROR; }
export function request() { throw NOT_SUPPORTED_ERROR; }
export function get() { throw NOT_SUPPORTED_ERROR; }

export const WebSocket = globalThis.WebSocket;
export const WebSocketStream = globalThis.WebSocketStream;

export default {
    METHODS, STATUS_CODES, maxHeaderSize,
    validateHeaderName, validateHeaderValue,
    Agent, globalAgent,
    ClientRequest, IncomingMessage,
    Server, ServerResponse,
    createServer, request, get,
    WebSocket, WebSocketStream,
};

File: crates/runtime/src/builtin/node_http_disabled.rs
#[rquickjs::module]
pub mod native_module {}

pub const NODE_HTTP_JS: &str = include_str!("node_http_disabled.js");
pub const HTTP_COMMON_JS: &str = include_str!("node_http_common.js");
pub const REEXPORT_JS: &str = r#"export * from 'node:http'; export { default } from 'node:http';"#;

pub const NODE_HTTP_SERVER_JS: &str = r#"
const msg = 'node:http server is not available (node-http feature is not enabled)';
function notAvailable() { throw new Error(msg); }
export const createServer = notAvailable;
export default { createServer };
"#;

pub const HTTP_AGENT_JS: &str = r#"
import { Agent } from 'node:http';
export { Agent };
export default Agent;
"#;
